boards/Kconfig

```diff
@@ -128,6 +128,11 @@ config QEMU_EXTRA_FLAGS
 	  to setup devices, for example to allocate interface for Zephyr
 	  GDBstub over serial with `-serial tcp:127.0.0.1:5678,server`
 
+config BOARD_SERIAL_BACKEND_CDC_ACM
+	bool "Board uses USB CDC ACM UART as serial backend"
```

Comment:
Segger RTT could be added in the same way later on.

Reply:
While using CDC ACM as the default backend for the boards without debug adapter, but with USB device controller and **bootloader** makes sense, what would be the case for "Segger RTT"? Is Segger RTT a subsystem that needs some configuration and how would that look like?

Reply:
We currently have this configuration available as a snippet (https://github.com/zephyrproject-rtos/zephyr/tree/main/snippets/rtt-console) which works fine for me. It was merely an example of another "serial" backend, that might be added like this in the future.

Reply:
Yes, and we know that there is also a cdc-acm-console snippet. But the requirements for this work are to make samples like "hello_world" work "out of the box" on boards with USB device controller and pre-flashed bootloader like nrf52840dongle. Are there similar use cases for boards with a Segger debug adapter? And if so, what would it look like? Is it worth adding support for this in boards/common and duplicate rtt-console snippet?

Reply:
I don't know if this need exists for in-tree boards, but for out-of-tree boards, it would certainly be a nice addition.

Reply:
> it would be against "General recommendations"

why would it be against "General recommendations" to have different serial backends ?

Enabling a different serial backend per default in upstream could be going against "General recommendations", but not having a choice which provides an option to select another backend.
A choice would furthermore ensure that only a single backend can be chosen at a time, like a user doing `-DCONFIG_SERIAL_BACKEND_FOO=y` will ensure that the CDC_ACM serial backend is not being selected.
For example when select RTT console.

Reply:
"Unless explicitly recommended otherwise by this section, leave peripherals and their drivers disabled by default." and "If available, configure and enable a serial output for the console..." IMO the wording only considers a real UART controller as a serial output.

>like a user doing -DCONFIG_SERIAL_BACKEND_FOO=y will ensure that the CDC_ACM serial backend is not being selected.
For example when select RTT console.

User doing -DCONFIG_SERIAL_BACKEND_FOO=y is completely wrong here, user capable to use build arguments should use `-DSNIPPET="rtt_console"` or `-DSNIPPET="cdc-acm-console"`. Backend configuration at the board level must be set to a single specific backend, if at all. No choices. If we want choices then snippets should be used and I can close this PR.

Reply:
I consider USB CDC to be a serial output device, and from my prior job as a vendor that had things like USB dongles added to zephyr, the company opinion was that they too are serial output devices and should be enabled by default

Reply:
> is completely wrong here, user capable to use build arguments should use -DSNIPPET="rtt_console" or -DSNIPPET="cdc-acm-console".

and what does `-DCONFIG_SERIAL_BACKEND_FOO` vs `-DSNIPPET=<foo>` selecting the backend have to do with the selecting being a choice or not ?

Even if this is configured through a snippet the Kconfig symbol should still be a choice.

Reply:
> and what does `-DCONFIG_SERIAL_BACKEND_FOO` vs `-DSNIPPET=<foo>` selecting the backend have to do with the selecting being a choice or not ?

`-DSNIPPET="cdc-acm-console"` selects logging serial backend, `-DCONFIG_BOARD_SERIAL_BACKEND_CDC_ACM=y` does not, it just enables USB device stack and applies few optional default settings. For this PR, **the board has already chosen the backend** using chosen property `zephyr,console` or `zephyr,shell-uart`, `boards/common/usb/Kconfig.cdc_acm_serial.defconfig` are dependencies for that choice.

Reply:
if this setting is not intended for the user to select, then the prompt should be removed.